README
WWW-Google-CustomSearch

The README  is  used  to  introduce the module and provide instructions on how to
install the module, any machine dependencies it may have (for example C compilers
and installed libraries) and any other information that should be provided before
the module is installed.

A  README  file  is required for CPAN modules since CPAN extracts the README file
from a module distribution so that people browsing the archive can use it to  get
an  idea  of  the  module's  uses.  It  is usually a good idea to provide version
information here so that people can decide whether fixes for the module are worth
downloading.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

Alternatively, to install with Module::Build, you can use the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After  installing,  you can  find documentation for this module with  the perldoc
command.

    perldoc WWW::Google::CustomSearch

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Google-CustomSearch

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/WWW-Google-CustomSearch

    CPAN Ratings
        http://cpanratings.perl.org/d/WWW-Google-CustomSearch

    Search CPAN
        http://search.cpan.org/dist/WWW-Google-CustomSearch/

LICENSE AND COPYRIGHT

Copyright (C) 2014 Mohammad S Anwar

This program is free software; you can redistribute it and/or modify it under the
terms of the the Artistic License (2.0).You may obtain a copy of the full license
at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying / distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you, you're nevertheless required to ensure that your Modified Version
complies with the requirements of this license.

This  license  does  not  grant you the right to use any trademark, service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to  make,  have made, use, offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by the  Package.  If you institute patent litigation
(including  a  cross-claim or  counterclaim)  against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF     MERCHANTABILITY,   FITNESS   FOR   A PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE,EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Changes
Revision history for WWW::Google::CustomSearch

0.36  Tue Jan 16 19:20:00 2018
      - Improved method exporting.

0.35  Mon Jun 12 14:10:00 2017
      - Updated dependency on WWW::Google::UserAgent v0.21.
      - Updated attribute definitions.
      - Updated unit test accordingly.

0.34  Wed Apr 12 18:15:00 2017
      - Tidied up test scripts and removed "-T" switch.

0.33  Mon Jan 30 11:30:00 2017
      - Fixed github issue #2 with regard to param key dateRestrict format. (thanks @electromike)

0.32  Sun Oct 04 15:50:00 2015
      - Upgraded to use WWW::Google::UserAgent v0.20.
      - Added dependency on Type::Tiny instead of Types::Standard.
      - Removed the upper limit restriction on the field 'start'.
      - Updated pod document about the restriction on the field 'start'.

0.31  Tue Sep 29 10:35:00 2015
      - Added dependency on Types::Standard.
      - Updated the way WWW::Google::UserAgent::DataType was being imported.
      - Upgraded to use WWW::Google::UserAgent v0.19.

0.30  Mon Sep 07 11:45:00 2015
      - Added link to the official Google API document.

0.29  Mon Aug 17 11:00:00 2015
      - Upgraded to use WWW::Google::UserAgent v0.18.

0.28  Thu Aug 06 12:00:00 2015
      - Upgraded to use WWW::Google::UserAgent v0.17.

0.27  Fri Mar 13 15:55:00 2015
      - Updated sample code in the section "SYNOPSIS".
      - Tidied up pod document in general.

0.26  THu Mar 12 11:40:00 2015
      - Updated section "SYNOPSIS" and added more sample code.
      - Tidied up pod document in general.

0.25  Sun Jan 18 19:35:00 2015
      - Upgraded to use WWW::Google::UserAgent v0.14 as it brings in 'https' support to
        patch issue RT #101571.
.
0.24  Mon Jan 12 14:40:18 2015
      - Upgraded to use WWW::Google::UserAgent v0.13.
      - Updated method search() in the pod document.

0.23  Mon Jan 12 12:30:18 2015
      - Added 'use 5.006;' line to the following test scripts:
        - t/01-new.t
        - t/02-search.t
        - t/meta-json.t
        - t/meta-yml.t

0.22  Sun Jan 11 21:45:18 2015
      - Upgraded to use WWW::Google::UserAgent v0.12.
      - Updated Makefile.PL to generate the key 'provides' for META file correctly.
      - Removed SIGNATURE file.
      - Added unit test to test META files.
      - Added LICENSE file.
      - Updated Copyright year information.

0.21  Thu Oct 16 20:33:29 2014
      - Fixed the syntax error in the code in the pod document of WWW::Google::CustomSearch package.

0.20  Mon Oct 13 20:27:20 2014
      - Updated contributor names in all pod document.
      - Minor changes to test script 00-load.t

0.19  Fri Oct 10 22:42:23 2014
      - Merged patch from David Kitcher-Jones about using URI for url encoding.
      - Added 'URI' as pre-requisite to the Makefile.PL file.
      - Updated pod documents (fixed code).

0.18  Fri Aug 29 10:10:00 2014
      - Added section "REPOSITORY" to the pod document.
      - Added key "META_MERGE" to Makefile.PL file.
      - Updated Copyright year details.

0.17  Tue Jul 29 15:08:15 2014
      - Made the pre-requisite WWW::Google::UserAgent v0.07

0.16  Tue Jul 29 00:20:19 2014
      - Made the pre-requisite WWW::Google::UserAgent v0.06

0.15  Mon Jul 28 14:16:19 2014
      - Made the pre-requisite WWW::Google::UserAgent v0.03

0.14  Sun Jul 27 11:27:00 2014
      - Upgraded to use WWW::Google::UserAgent v0.02.
      - Removed Build.PL file.

0.13  Sat Jul 19 04:00:00 2014
      - Restructured code to use Moo.
      - Making use of common library WWW::Google::UserAgent.

0.12  Sat Mar 29 08:27:13 2014
      - Added dependency check for LWP::Protocol::https v6.04 (RT: 94194).

0.11  Sat Mar 01 11:12:13 2014
      - Added link to the Google JSON/Atom Custom Search API in the pod.

0.10  Sun Feb 02 17:15:10 2014
      - Fixed the issue raised by JACINTA (RT: 92596).

0.09  Fri Sep 14 08:55:10 2012
      - Fixed the warning when no items found.
      - Explicitly set the defaults.
      - Hide the 'alt' attribute from user as it is no longer relevant.

0.08  Fri Sep 14 06:02:22 2012
      - Added XML::Simple dependency check.

0.07  Thu Sep 13 15:00:00 2012
      - Simplied the process of fetching search result.
      - Fixed the handling of 'alt' attribute.

0.06  Wed May 02 11:15:00 2012
      - Updated pod about how the query parameters can be specified.

0.05  Tue Nov 15 11:12:13 2011
      - Fixed the issued raised by WILLERT (RT: 72417).

0.04  Mon Sep 12 17:18:19 2011
      - Fixed the checking of fields 'num' and 'start'.
      - Added unit test to keep an eye.

0.03  Mon Sep 12 14:15:16 2011
      - Handle default settings correctly.

0.02  Mon Aug 08 12:44:11 2011
      - Made Readonly v1.03 as minimum requirement.
      - Prepared the distribution using the Module::Signature v0.68.

0.01  Wed May 18 19:39:01 2011
      - First version, released.

lib/WWW/Google/CustomSearch/Item.pm
package WWW::Google::CustomSearch::Item;

$WWW::Google::CustomSearch::Item::VERSION   = '0.36';
$WWW::Google::CustomSearch::Item::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

WWW::Google::CustomSearch::Item - Placeholder for Google JSON/Atom Custom Search Item.

=head1 VERSION

Version 0.36

=cut

use 5.006;
use Data::Dumper;

use Moo;
use namespace::clean;

has 'kind'             => (is => 'ro');
has 'link'             => (is => 'ro');
has 'displayLink'      => (is => 'ro');
has 'snippet'          => (is => 'ro');
has 'htmlSnippet'      => (is => 'ro');
has 'cacheId'          => (is => 'ro');
has 'formattedUrl'     => (is => 'ro');
has 'htmlFormattedUrl' => (is => 'ro');
has 'title'            => (is => 'ro');
has 'htmlTitle'        => (is => 'ro');

=head1 DESCRIPTION

Provides  the  interface  to the individual search page result items based on the
search criteria.

=head1 METHODS

=head2 kind()

Returns the 'kind' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item (@{$result->items}) {
        print "Item Kind: ", $item->kind, "\n" if defined $item->kind;
    }

=head2 link()

Returns the 'link' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item (@{$result->items}) {
        print "Item Link: ", $item->link, "\n" if defined $item->link;
    }

=head2 displayLink()

Returns the 'displayLink' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item (@{$result->items}) {
        print "Item Display Link: ", $item->displayLink, "\n" if defined $item->displayLink;
    }

=head2 snippet()

Returns the 'snippet' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item (@{$result->items}) {
        print "Item Snippet: ", $item->snippet, "\n" if defined $item->snippet;
    }

=head2 htmlSnippet()

Returns the 'htmlSnippet' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item (@{$result->items}) {
        print "Item HTML Snippet: ", $item->htmlSnippet, "\n" if defined $item->htmlSnippet;
    }

=head2 cacheId()

Returns the 'cacheId' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item (@{$result->items}) {
        print "Item Cache Id: ", $item->cacheId, "\n" if defined $item->cacheId;
    }

=head2 formattedUrl()

Returns the 'formattedUrl' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item (@{$result->items}) {
        print "Item Formatted URL: ", $item->formattedUrl, "\n" if defined $item->formattedUrl;
    }

=head2 htmlFormattedUrl()

Returns the 'htmlFormattedUrl' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item (@{$result->items}) {
        print "Item HTML Formatted URL: ", $item->htmlFormattedUrl, "\n" if defined $item->htmlFormattedUrl;
    }

=head2 title()

Returns the 'title' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item (@{$result->items}) {
        print "Item Title: ", $item->title, "\n" if defined $item->title;
    }

=head2 htmlTitle()

Returns the 'htmlTitle' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    foreach my $item (@{$result->items}) {
        print "Item HTML Title: ", $item->htmlTitle, "\n" if defined $item->htmlTitle;
    }

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/WWW-Google-CustomSearch>

=head1 CONTRIBUTORS

David Kitcher-Jones (m4ddav3)

=head1 BUGS

Please  report  any  bugs or feature requests to C<bug-www-google-customsearch at
rt.cpan.org>, or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Google-CustomSearch>.
I will be notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Google::CustomSearch::Item

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Google-CustomSearch>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Google-CustomSearch>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Google-CustomSearch>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Google-CustomSearch/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2015 Mohammad S Anwar.

This program  is  free software; you can redistribute it and / or modify it under
the  terms  of the the Artistic License (2.0). You may obtain  a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of WWW::Google::CustomSearch::Item


lib/WWW/Google/CustomSearch/Page.pm
package WWW::Google::CustomSearch::Page;

$WWW::Google::CustomSearch::Page::VERSION   = '0.36';
$WWW::Google::CustomSearch::Page::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

WWW::Google::CustomSearch::Page - Placeholder for Google JSON/Atom Custom Search Page.

=head1 VERSION

Version 0.36

=cut

use 5.006;
use Data::Dumper;

use Moo;
use namespace::clean;

has  'api_key'        => (is => 'ro', required => 1);
has  'cx'             => (is => 'ro', required => 1);
has  'safe'           => (is => 'ro', default => sub { return 'off' });
has  'count'          => (is => 'ro');
has  'searchTerms'    => (is => 'ro', required => 1);
has  'inputEncoding'  => (is => 'ro', required => 1);
has  'startIndex'     => (is => 'ro', default  => sub { return 1 });
has  'title'          => (is => 'ro', required => 1);
has  'totalResults'   => (is => 'ro', required => 1);
has  'outputEncoding' => (is => 'ro', required => 1);

=head1 DESCRIPTION

Provides the interface to the individual search page based on the search criteria.

=head1 METHODS

=head2 safe()

Returns the safety level of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Safety Level: ", $page->safe, "\n";

=head2 count()

Returns the 'count' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Page count: ", $page->count, "\n";

=head2 searchTerms()

Returns the 'searchTerms' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Search Terms: ", $page->searchTerms, "\n";

=head2 startIndex()

Returns the 'startIndex' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Start Index: ", $page->startIndex, "\n";

=head2 title()

Returns the 'title' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Title: ", $page->title, "\n";

=head2 totalResults()

Returns the 'totalResults' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Total Results: ", $page->totalResults, "\n";

=head2 inputEncoding()

Returns the 'inputEncoding' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Input Encoding: ", $page->inputEncoding, "\n";

=head2 outputEncoding()

Returns the 'outputEncoding' attribute of the search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->nextPage;
    print "Output Encoding: ", $page->outputEncoding, "\n";

=head2 fetch()

Perform a fresh search based on the previous input data and returns L<WWW::Google::CustomSearch::Result>
object.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key    = 'Your_API_Key';
    my $cx         = 'Search_Engine_Identifier';
    my $engine     = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result     = $engine->search("Google");
    my $page       = $result->nextPage;
    my $nextResult = $page->fetch;

=cut

sub fetch {
    my ($self) = @_;

    my $engine = WWW::Google::CustomSearch->new({ api_key => $self->api_key, cx => $self->cx, start => $self->startIndex });
    return $engine->search($self->searchTerms);
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/WWW-Google-CustomSearch>

=head1 CONTRIBUTORS

David Kitcher-Jones (m4ddav3)

=head1 BUGS

Please  report any bugs or feature requests  to  C<bug-www-google-customsearch at
rt.cpan.org>, or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Google-CustomSearch>.
I will be notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Google::CustomSearch::Page

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Google-CustomSearch>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Google-CustomSearch>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Google-CustomSearch>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Google-CustomSearch/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2015 Mohammad S Anwar.

This program  is  free software; you can redistribute it and / or modify it under
the  terms  of the the Artistic License (2.0). You may obtain  a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of WWW::Google::CustomSearch::Page


lib/WWW/Google/CustomSearch/Request.pm
package WWW::Google::CustomSearch::Request;

$WWW::Google::CustomSearch::Request::VERSION   = '0.36';
$WWW::Google::CustomSearch::Request::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

WWW::Google::CustomSearch::Request - Placeholder for Google JSON/Atom Custom Search Request.

=head1 VERSION

Version 0.36

=cut

use 5.006;
use Data::Dumper;
use WWW::Google::CustomSearch::Page;

use Moo;
use namespace::clean;

has 'page'    => (is => 'ro');
has 'api_key' => (is => 'ro', required => 1);

sub BUILD {
    my ($self, $args) = @_;

    $args->{'api_key'} = $self->api_key;
    $self->{page} = WWW::Google::CustomSearch::Page->new($args);
}

=head1 DESCRIPTION

Provides the interface to the search request used last time.

=head1 METHOD

=head2 fetch()

Returns the L<WWW::Google::CustomSearch::Result> object base on the criteria used in the  last
search.

    use strict; use warnings;
    use WWW::Google::CustomSearch;

    my $api_key = 'Your_API_Key';
    my $cx      = 'Search_Engine_Identifier';
    my $engine  = WWW::Google::CustomSearch->new(api_key => $api_key, cx => $cx);
    my $result  = $engine->search("Google");
    my $page    = $result->request;
    my $another = $page->fetch;

=cut

sub fetch {
    my ($self) = @_;

    return $self->{page}->fetch;
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/Manwar/WWW-Google-CustomSearch>

=head1 CONTRIBUTORS

David Kitcher-Jones (m4ddav3)

=head1 BUGS

Please report  any bugs or feature  requests to  C<bug-www-google-customsearch at
rt.cpan.org>, or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Google-CustomSearch>.
I will be notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Google::CustomSearch::Request

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Google-CustomSearch>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Google-CustomSearch>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Google-CustomSearch>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Google-CustomSearch/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2015 Mohammad S Anwar.

This program  is  free software; you can redistribute it and / or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a  copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of WWW::Google::CustomSearch::Request


Makefile.PL
#!/usr/bin/perl

use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'WWW::Google::CustomSearch',
    AUTHOR             => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    VERSION_FROM       => 'lib/WWW/Google/CustomSearch.pm',
    ABSTRACT_FROM      => 'lib/WWW/Google/CustomSearch.pm',
    LICENSE            => 'artistic_2',
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES   => {
        'Test::More' => 0,
    },
    PREREQ_PM  => {
        'Moo'                    => '2.000000',
        'namespace::clean'       => '0.24',
        'JSON'                   => '2.53',
        'URI'                    => '1.65',
        'WWW::Google::UserAgent' => '0.21',
        'Type::Tiny'             => '1.000005'
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'WWW-Google-CustomSearch-*' },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        provides    => {
            'WWW::Google::CustomSearch'          => { file => 'lib/WWW/Google/CustomSearch.pm',         version => '0.36' },
            'WWW::Google::CustomSearch::Item'    => { file => 'lib/WWW/Google/CustomSearch/Item.pm',    version => '0.36' },
            'WWW::Google::CustomSearch::Page'    => { file => 'lib/WWW/Google/CustomSearch/Page.pm',    version => '0.36' },
            'WWW::Google::CustomSearch::Params'  => { file => 'lib/WWW/Google/CustomSearch/Params.pm',  version => '0.36' },
            'WWW::Google::CustomSearch::Request' => { file => 'lib/WWW/Google/CustomSearch/Request.pm', version => '0.36' },
            'WWW::Google::CustomSearch::Result'  => { file => 'lib/WWW/Google/CustomSearch/Result.pm',  version => '0.36' },
        },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/manwar/WWW-Google-CustomSearch.git',
                web  => 'https://github.com/manwar/WWW-Google-CustomSearch',
            },
        }})
     : ()
    ),
);
